src/v/coproc/partition.cc

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "coproc/partition.h"
+
+#include "storage/log.h"
+
+namespace coproc {
+/// Not neccessary to hold gate open until all processing completes, as that can
+/// be done by invoking the write within the context of the \ref
+/// non_replicable_partition gate. This class just makes the operation
+/// cancellable
+safe_shutdown_appender::safe_shutdown_appender(
+  ss::gate& g, storage::log_appender&& appender) noexcept
+  : _gate(g)
+  , _appender(std::move(appender)) {}
+
+ss::future<ss::stop_iteration>
+safe_shutdown_appender::operator()(model::record_batch& rb) {
+    if (_gate.is_closed()) {
+        return ss::make_ready_future<ss::stop_iteration>(
+          ss::stop_iteration::yes);
+    }
+    return ss::do_with(std::move(rb), [this](model::record_batch& rb) {
+        return ss::with_gate(
+          _gate, [this, &rb] { return _appender.operator()(rb); });
+    });
+}
+
+ss::future<storage::append_result> safe_shutdown_appender::end_of_stream() {
+    return _appender.end_of_stream();
```

Comment:
why don't you check the gate here?

Reply:
Resolving these comments as i am moving away from this solution